### ISSUE TITLE
bug in spec generation when objects are contained and references itself (recursion)

### ISSUE THREAD

**Issue by mortias** [swagger-api/swagger-core#1178]
Basically, the spec generates fine when the object references itself or the contained references the container. However, the reference is missing (in the spec) when the contained references itself.

We've identified this problem here: https://github.com/kongchen/swagger-maven-plugin/issues/177
it should have been fixed https://github.com/swagger-api/swagger-core/issues/1131 but it isn't i think

I have a simple get rest method
```
@Api(description = "Operations about pets")
@Path("/test")
public class TestResource {
    @GET
    @Path("menudata")
    @Produces({"application/json"})
    @ApiOperation(httpMethod = "GET", value = "Get the menu", response = MenuData.class)
    public MenuData getMenu() {
        MenuData root = new MenuData();
        return root;
    }
}
```
My MenuData Class that holds a menu
```
package api;
import java.util.List;
public class MenuData {
   public List<MenuItem> menudata;
}
```
My MenuItem Class which is recursive, public List<MenuItem> sub;
```
public class MenuItem {

    public String description;
    public String title;

    public List<MenuItem> sub;

    public MenuItem(String title, String description) {
        this.title = title;
        this.description = description;
        this.sub = new ArrayList<MenuItem>();
    }

}
```
This then generates in the MenuItem definition an array without any reference.. 
So it seems that the recursion isn't solved yet i guess 
       "sub" : {
          "type" : "array"
        }
```
{
  "swagger" : "2.0",
  "info" : {
    "description" : "This is a sample.",
    "version" : "v1",
    "title" : "Swagger Maven Plugin Sample"
  },
  "host" : "localhost:9998",
  "basePath" : "/api",
  "schemes" : [ "http" ],
  "paths" : {
    "/test/menudata" : {
      "get" : {
        "summary" : "Get the menu",
        "description" : "",
        "operationId" : "getMenu",
        "produces" : [ "application/json" ],
        "responses" : {
          "200" : {
            "description" : "successful operation",
            "schema" : {
              "$ref" : "#/definitions/MenuData"
            }
          }
        }
      }
    }
  },
  "definitions" : {
    "MenuData" : {
      "type" : "object",
      "properties" : {
        "menudata" : {
         "type" : "array",
          "items" : {
            "$ref" : "#/definitions/MenuItem"
          }
        }
      }
    },
    "MenuItem" : {
      "type" : "object",
      "properties" : {
        "description" : {
          "type" : "string"
        },
        "title" : {
          "type" : "string"
        },
       "sub" : {
          "type" : "array"
        }
      }
    }
  }
}
```

**Comment by mortias** [comment]
Note: if i change the 
```
public List<MenuItem> sub; 
```
to another class type, it's fine
```
public List<MenuNode> sub;
```
result: 
```
"definitions" : {
    "MenuData" : {
      "type" : "object",
      "properties" : {
        "menudata" : {
          "type" : "array",
          "items" : {
            "$ref" : "#/definitions/MenuItem"
          }
        }
      }
    },
    "MenuItem" : {
      "type" : "object",
      "properties" : {
        "description" : {
          "type" : "string"
        },
        "title" : {
          "type" : "string"
        },
        "sub" : {
          "type" : "array",
          "items" : {
            "$ref" : "#/definitions/MenuNode"
          }
        }
      }
    },
    "MenuNode" : {
      "type" : "object",
      "properties" : {
        "node" : {
          "type" : "string"
        }
      }
    }
```

**Comment by mortias** [comment]
If i only return the menuItem (without being used in another class like the menuData, its also fine)
```
  @GET
    @Path("menuitems")
    @Produces({"application/json"})
    @ApiOperation(httpMethod = "GET", value = "Get the menu", response = MenuItem.class)
    public MenuItem getMenu() {
        MenuItem root = new MenuItem("a","b");
        return root;
    }
```
```
public class MenuItem {

    public String description;
    public String title;

    public List<MenuItem> sub;

    public MenuItem(String title, String description) {
        this.title = title;
        this.description = description;
        this.sub = new ArrayList<MenuItem>();
    }

}
```
```
 "paths" : {
    "/test/menudata" : {
      "get" : {
        "summary" : "Get the menu",
        "description" : "",
        "operationId" : "getMenu",
        "produces" : [ "application/json" ],
        "responses" : {
          "200" : {
            "description" : "successful operation",
            "schema" : {
              "$ref" : "#/definitions/MenuItem"
            }
          }
        }
      }
    }
  },
  "definitions" : {
    "MenuItem" : {
      "type" : "object",
      "properties" : {
        "description" : {
          "type" : "string"
        },
        "title" : {
          "type" : "string"
        },
        "sub" : {
          "type" : "array",
          "items" : {
            "$ref" : "#/definitions/MenuItem"
          }
        }
      }
    }
  }
}
```

**Comment by mortias** [comment]
Fyi

```
  <!-- swagger -->
        <dependency>
            <groupId>io.swagger</groupId>
            <artifactId>swagger-core</artifactId>
            <version>1.5.0</version>
        </dependency>

       <build>
        <plugins>
            <plugin>
                <groupId>com.github.kongchen</groupId>
                <artifactId>swagger-maven-plugin</artifactId>
                <version>3.1.0</version>
                <configuration>
                    <apiSources>
                        <apiSource>
                            <springmvc>false</springmvc>
                            <schemes>http</schemes>
                            <host>localhost:9998</host>
                            <basePath>/api</basePath>
                            <info>
                                <title>Swagger Maven Plugin Sample</title>
                                <version>v1</version>
                                <description>
                                    This is a sample.
                                </description>
                            </info>
                            <locations>rest</locations>
                            <swaggerDirectory>${project.basedir}/target/swagger-json/</swaggerDirectory>
                        </apiSource>
                    </apiSources>
                </configuration>
                <executions>
                    <execution>
                        <phase>compile</phase>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
```

**Comment by mortias** [comment]
So there is good and bad news, 
the bad news is that there is some sort of problem and it looses track of the right reference probably,..

but the good news is, i found a "workaround", so it's not blocking, however it's not what i want..
if i specify the DataType i get an array with an object definition so swagger.ui doesn't block..
however the "type" : "object" should point to the definition of "MenuItem"..

```
    @ApiModelProperty(required = true, dataType="java.util.ArrayList")
    public ArrayList<MenuItem> sub;
```

results in 
``` 
  "MenuItem" : {
      "type" : "object",
      "required" : [ "sub" ],
      "properties" : {
        "description" : {
          "type" : "string"
        },
        "title" : {
          "type" : "string"
        },
        "sub" : {
          "type" : "array",
          "items" : {
            "type" : "object"
          }
        }
      },
      "description" : "MenuItem"
    }
```

**Comment by webron** [comment]
@mortias - you said it is not working with the maven plugin, but have you tested the real time generation with swagger-core?

**Comment by tomtit** [comment]
@webron we have this PR #1163 for the issue.

**Comment by webron** [comment]
@mortias - can you try with the snapshot version?

**Comment by blop** [comment]
We have the same bug, compiling 1.5.1-SNAPSHOT from branch Develop_2.0 to test ;-)

**Comment by blop** [comment]
I confirm this is fixed in upcoming 1.5.1-SNAPSHOT version.
Thanks

**Comment by webron** [comment]
Thanks for the confirmation. Closed as fixed in develop_2.0.

**Comment by mortias** [comment]
sorry for the late reply but i also confirm it in my case

**Comment by mortias** [comment]
Please re-open the issue and test it with my simple case above..

**Comment by webron** [comment]
@mortias - just tested it, works as expected.

**Comment by mortias** [comment]
Hmm... i 'll test it again then :)

**Comment by xinmeng1** [comment]
The bug is fixed on swagger-maven-plugin >3.1.2, I have test it is OK on the swagger-maven-plugin =3.1.2 & swagger-core 1.5.9.
